test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetEndToEndTests/NuGetUITestCase.cs

```diff
@@ -735,5 +735,96 @@ public void UninstallDeprecatedPackageFromUI(ProjectTemplate projectTemplate, st
             // Assert
             CommonUtility.AssertUninstalledPackageByProjectType(VisualStudio, projectTemplate, project, packageName, Logger);
         }
+
+        [TestMethod]
+        [Timeout(DefaultTimeout)]
+        public void InstallTopLevelPackageHavingTransitiveFromUI()
+        {
+            // Arrange
+            EnsureVisualStudioHost();
+            var dte = VisualStudio.Dte;
+            var solutionService = VisualStudio.Get<SolutionService>();
+            solutionService.CreateEmptySolution();
+            var project = solutionService.AddProject(ProjectLanguage.CSharp, ProjectTemplate.NetCoreClassLib, "TestProject");
+            VisualStudio.ClearOutputWindow();
+            solutionService.SaveAll();
+
+            // Act
+            CommonUtility.OpenNuGetPackageManagerWithDte(VisualStudio, Logger);
+            var nugetTestService = GetNuGetTestService();
+            var uiwindow = nugetTestService.GetUIWindowfromProject(project);
+            uiwindow.InstallPackageFromUI("log4net", "2.0.15");
```

Comment:
I'm guessing this package is coming from nuget.org?

I think @zivkan's feedback is to generate a package in a local feed and use that instead of using live packages.

Reply:
But the aim of these three test cases is testing the transitive package. Creating a package in a local feed means the normal package, correct? We thought Andy just means the "Moq" is blocked by internal compliance tooling, so we changed it to another.

Reply:
I've provided feedback in the past that I don't believe tests should use nuget.org, but I feel like this suggestion often gets ignored (across the whole team, not just regarding Apex tests), so I've kind of given up explicitly asking for that.  But I do wish we'd stop using nuget.org packages so that we minimize issues when new vulnerabilities are disclosed, new package versions published (perhaps no longer compatible with the test project TFM, and the test was installing the latest version), and other causes for test failures or compliance issues.

Many of our restore tests use generated packages, including some with dependencies:
https://github.com/NuGet/NuGet.Client/blob/5ce332fb834feea1f6c974d17456c8c9d5504e09/test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/RestoreCommandTests.cs#L655-L662

Then use the helper method to create the packages in the package source directory: https://github.com/NuGet/NuGet.Client/blob/5ce332fb834feea1f6c974d17456c8c9d5504e09/test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/RestoreCommandTests.cs#L668-L672

However, `SimpleTestPathContext` assumes the tests are running within the NuGet.Client repo (it searches for the repo root, then creates subdirectories under `{repo root}\.test\`). But Apex tests are compiled on one machine and run on a different machine without the repo. Therefore, some changes need to be made to the test helper methods, and `solutionService.CreateEmptySolution();` needs to be investigated to be created in the solution subdirectory that `SimpleTestPathContext` creates, rather than whatever location it currently uses (temp directory?).  Anyway, all this is one-time work, and then we'll have helper methods that allow us to generate packages for Apex tests easily.

We still need to avoid using names that match packages with security vulnerabilities (so don't call the test package `moq` or `jquery`, etc), as the compliance tool doesn't know the difference. But using obviously test package names is unlikely to

Reply:
> I've provided feedback in the past that I don't believe tests should use nuget.org, but I feel like this suggestion often gets ignored (across the whole team, not just regarding Apex tests), so I've kind of given up explicitly asking for that. But I do wish we'd stop using nuget.org packages so that we minimize issues when new vulnerabilities are disclosed, new package versions published (perhaps no longer compatible with the test project TFM, and the test was installing the latest version), and other causes for test failures or compliance issues.
> 
> Many of our restore tests use generated packages, including some with dependencies:
> 
> https://github.com/NuGet/NuGet.Client/blob/5ce332fb834feea1f6c974d17456c8c9d5504e09/test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/RestoreCommandTests.cs#L655-L662
> 
> Then use the helper method to create the packages in the package source directory:
> 
> https://github.com/NuGet/NuGet.Client/blob/5ce332fb834feea1f6c974d17456c8c9d5504e09/test/NuGet.Clients.FuncTests/NuGet.CommandLine.FuncTest/Commands/RestoreCommandTests.cs#L668-L672
> 
> However, `SimpleTestPathContext` assumes the tests are running within the NuGet.Client repo (it searches for the repo root, then creates subdirectories under `{repo root}\.test\`). But Apex tests are compiled on one machine and run on a different machine without the repo. Therefore, some changes need to be made to the test helper methods, and `solutionService.CreateEmptySolution();` needs to be investigated to be created in the solution subdirectory that `SimpleTestPathContext` creates, rather than whatever location it currently uses (temp directory?). Anyway, all this is one-time work, and then we'll have helper methods that allow us to generate packages for Apex tests easily.
> 
> We still need to avoid using names that match packages with security vulnerabilities (so don't call the test package `moq` or `jquery`, etc), as the compliance tool doesn't know the difference. But using obviously test package names is unlikely to

Hi @zivkan & @nkolev92, thanks for your suggestion, I have change the package to local package in cases, could you please help to review it, thanks a lot!